src/mapgen/mapgen_valleys.cpp

```diff
@@ -529,48 +425,37 @@ int MapgenValleys::generateTerrain()
 			index_3d += ystride;
 		}
 
-		if (heightmap[index_2d] == -MAX_MAP_GENERATION_LIMIT) {
-			s16 surface_y_int = myround(surface_y);
-
-			if (surface_y_int > node_max.Y + 1 ||
-					surface_y_int < node_min.Y - 1) {
-				// If surface_y is outside the chunk, it's good enough
-				heightmap[index_2d] = surface_y_int;
-			} else {
-				// If the ground is outside of this chunk, but surface_y is
-				// within the chunk, give a value outside.
-				heightmap[index_2d] = node_min.Y - 2;
-			}
-		}
-
 		// Optionally increase humidity around rivers
 		if (spflags & MGVALLEYS_HUMID_RIVERS) {
+			// Compensate to avoid increasing average humidity
+			m_bgen->humidmap[index_2d] *= 0.8f;
 			// Ground height ignoring riverbeds
-			float t_alt = std::fmax(noise_rivers->result[index_2d],
-				(float)heightmap[index_2d]);
-			float water_depth = (t_alt - river_y) / humidity_dropoff;
+			float t_alt = std::fmax(base, (float)column_max_y);
+			float water_depth = (t_alt - base) / 4.0f;
 			m_bgen->humidmap[index_2d] *=
 				1.0f + std::pow(0.5f, std::fmax(water_depth, 1.0f));
 		}
 
 		// Optionally decrease humidity with altitude
 		if (spflags & MGVALLEYS_ALT_DRY) {
 			// Ground height ignoring riverbeds
-			float t_alt = std::fmax(noise_rivers->result[index_2d],
-				(float)heightmap[index_2d]);
-			if (t_alt > 0.0f)
+			float t_alt = std::fmax(base, (float)column_max_y);
+			// Only decrease above water_level
+			if (t_alt > water_level)
 				m_bgen->humidmap[index_2d] -=
-					alt_to_humid * t_alt / altitude_chill;
+					(t_alt - water_level) * 10.0f / altitude_chill;
 		}
 
 		// Optionally decrease heat with altitude
 		if (spflags & MGVALLEYS_ALT_CHILL) {
+			// Compensate to avoid reducing the average heat
+			m_bgen->heatmap[index_2d] += 5.0f;
```

Comment:
Is this based on any calculation? Looks a lot like a magic value.

Reply:
Probably another arbitrary value, that was already in duane's code: https://github.com/minetest/minetest/blob/1511086/src/mapgen/mapgen_valleys.cpp#L302.
It's a compensation because the temperature is decreased with elevation, but since the temperature varies linearly and the elevation has no strict maximum or known mean, I don't see how we can define any constant.

Reply:
Yes just a value chosen through tuning, was also present in previous mgvalleys code (but in a different place).